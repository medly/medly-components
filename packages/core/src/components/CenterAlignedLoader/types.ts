import { HTMLProps } from '@medly-components/utils';
import type { ReactNode } from 'react';

export interface CenterAlignedProps extends HTMLProps<HTMLDivElement> {
    /** Set it true to show the box with text and loader */
    withLoadingBox?: boolean;
    /** Set it true to show overlay */
    withOverlay?: boolean;
    /** Use it to set the position of a wrapper component */
    position?: 'static' | 'absolute' | 'fixed' | 'relative' | 'sticky' | 'initial' | 'inherit';
}

export interface CenterAlignedLoaderProps extends CenterAlignedProps {
    /** Use any react component as a loader */
    loader?: ReactNode;
}
