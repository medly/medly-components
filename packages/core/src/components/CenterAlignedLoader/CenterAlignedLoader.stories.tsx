import styled from 'styled-components';

export const RelativeDiv = styled.div`
    position: relative;
`;
