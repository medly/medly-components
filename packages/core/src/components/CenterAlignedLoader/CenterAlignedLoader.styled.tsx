import { rgba } from 'polished';
import styled, { keyframes } from 'styled-components';
import Text from '../Text';
import { CenterAlignedProps } from './types';

const fadesIn = keyframes`
  from {
    transform: translateY(0.8rem);
    opacity: 0;
  }

  to {
    transform: translateY(0);
    opacity: 1;
  }
`;

export const CenterAligned = styled('div')<CenterAlignedProps>`
    height: 100%;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 1001;
    position: ${({ position }) => position};
    background-color: ${({ withOverlay, theme }) => withOverlay && rgba(theme.colors.white, 0.75)};
    & > * {
        animation: ${fadesIn} 300ms ${({ withLoadingBox }) => (withLoadingBox ? 'ease-in-out' : 'ease-in')};
    }
`;

export const LoadingBox = styled('div')`
    box-sizing: border-box;
    background-color: ${({ theme }) => theme.colors.white};
    height: 14.4rem;
    min-width: 14.4rem;
    padding: 0 1rem;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    box-shadow: 0 0.4rem 1.6rem ${({ theme }) => rgba(theme.colors.grey[600], 0.2)};
    border-radius: 1rem;
    ${Text.Style} {
        margin-top: 1.4rem;
        color: ${({ theme }) => theme.colors.grey[700]};
    }
`;
