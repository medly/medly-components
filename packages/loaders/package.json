{
    "name": "@medly-components/loaders",
    "version": "1.2.33",
    "sideEffects": false,
    "main": "dist/cjs/index.js",
    "module": "dist/es/index.js",
    "types": "dist/es/index.d.ts",
    "files": [
        "dist"
    ],
    "publishConfig": {
        "access": "public"
    },
    "license": "MIT",
    "peerDependencies": {
        "react": "16.x || 17.x",
        "react-dom": "16.x || 17.x",
        "styled-components": "^5.1.0"
    },
    "dependencies": {
        "@medly-components/theme": "2.19.0",
        "@medly-components/utils": "1.14.11"
    }
}
