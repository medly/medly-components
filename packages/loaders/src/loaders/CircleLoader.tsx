import { WithStyle } from '@medly-components/utils';
import React, { FC } from 'react';
import CircleLoaderSvg from '../assets/CircleLoader.svg';
import SvgLoader from '../SvgLoader';
import { SvgLoaderProps } from '../SvgLoader/types';

const Component: FC<SvgLoaderProps> = React.memo(props => (
    <SvgLoader {...props}>
        <CircleLoaderSvg {...props} width="1em" height="1em" />
    </SvgLoader>
));

Component.displayName = 'CircleLoader';
export const CircleLoader: FC<SvgLoaderProps> & WithStyle = Object.assign(Component, { Style: SvgLoader });
