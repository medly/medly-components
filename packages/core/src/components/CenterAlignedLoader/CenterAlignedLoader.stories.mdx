import { CenterAlignedLoader } from './CenterAlignedLoader.tsx';
import { Text } from '@medly-components/core';
import { ConcentricCircleLoader } from '@medly-components/loaders';
import { Preview, Meta, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';
import { RelativeDiv } from './CenterAlignedLoader.stories.tsx';

<Meta title="Core" component={CenterAlignedLoader} parameters={{ jest: ['CenterAlignedLoader.test.tsx'] }} />

# CenterAlignedLoader

`CenterAlignedLoader` displays a loading state while the main component mounts or the data loads. This component takes the width and height of the parent and then renders the loader in the center. 
Check out all the available loaders in [Medly Components](https://medly.github.io/medly-components/?path=/story/loaders--loaders).

<Preview withToolbar>
    <Story name="CenterAlignedLoader" decorators={[withKnobs]}>
        <CenterAlignedLoader withLoadingBox={boolean('With loading box', true)} withOverlay={boolean('With overlay', true)} />
    </Story>
</Preview>

## Example with different loader

To pass any React component as a loader, use: 

`<CenterAlignedLoader withLoadingBox loader={<ConcentricCircleLoader size="XS" />} />`

<Preview withToolbar>
    <CenterAlignedLoader withLoadingBox loader={<ConcentricCircleLoader size="XS" />} />
</Preview>

## Example with absolute position

To display a loader on top of another component, set the loader position to `absolute` and the parent position to `relative`.Â 

<Preview withToolbar>
    <RelativeDiv>
        <CenterAlignedLoader withOverlay position="absolute" />
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever
        since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. Ithas survived not only
        five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It waspopularised in the 1960s with the
        release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktoppublishing software like Aldus PageMaker including
        versions of Lorem Ipsum
    </RelativeDiv>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={CenterAlignedLoader} />
