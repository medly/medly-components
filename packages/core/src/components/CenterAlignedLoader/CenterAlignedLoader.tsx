import { WithStyle } from '@medly-components/utils';
import type { FC } from 'react';
import { memo, useMemo } from 'react';
import Text from '../Text';
import { CenterAligned, LoadingBox } from './CenterAlignedLoader.styled';
import Loader from './CircleLoader.svg';
import { CenterAlignedLoaderProps } from './types';

const Component: FC<CenterAlignedLoaderProps> = memo(({ loader, withLoadingBox, loadingMessage, ...restProps }) => {
    const children = useMemo(
        () =>
            withLoadingBox ? (
                <LoadingBox>
                    {loader}
                    <Text>{loadingMessage ? loadingMessage : 'Loading...'}</Text>
                </LoadingBox>
            ) : (
                <>{loader}</>
            ),
        [withLoadingBox, loadingMessage, loader]
    );
    return (
        <CenterAligned withLoadingBox={withLoadingBox} {...restProps}>
            {children}
        </CenterAligned>
    );
});
Component.displayName = 'CenterAlignedLoader';
Component.defaultProps = { loader: <Loader />, withLoadingBox: false, loadingMessage: '', withOverlay: false, position: 'relative' };
export const CenterAlignedLoader: FC<CenterAlignedLoaderProps> & WithStyle = Object.assign(Component, { Style: CenterAligned });
