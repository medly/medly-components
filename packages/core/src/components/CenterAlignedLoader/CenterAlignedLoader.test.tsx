import { ConcentricCircleLoader } from '@medly-components/loaders';
import { render } from '@test-utils';
import { CenterAlignedLoader } from './CenterAlignedLoader';

describe('Center Aligned Loader', () => {
    it('should render properly with default props', () => {
        const { container } = render(<CenterAlignedLoader />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly with all props given', () => {
        const { container } = render(
            <CenterAlignedLoader withLoadingBox withOverlay position="absolute" loader={<ConcentricCircleLoader size="XS" />} />
        );
        expect(container).toMatchSnapshot();
    });
});
