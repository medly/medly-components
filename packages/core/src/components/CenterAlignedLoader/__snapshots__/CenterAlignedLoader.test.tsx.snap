// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Center Aligned Loader should render properly with all props given 1`] = `
.c2 {
  font-size: 48px;
}

.c2 * {
  fill: #3872D2;
}

.c4 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c0 {
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 1001;
  position: absolute;
  background-color: rgba(255,255,255,0.75);
}

.c0 > * {
  -webkit-animation: hHgZmL 300ms ease-in-out;
  animation: hHgZmL 300ms ease-in-out;
}

.c1 {
  box-sizing: border-box;
  background-color: #ffffff;
  height: 14.4rem;
  width: 14.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-shadow: 0 0.4rem 1.6rem rgba(96,120,144,0.2);
  border-radius: 1rem;
}

.c1 .c3 {
  margin-top: 1.4rem;
  color: #546A7F;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <svg
        class="c2"
        height="1em"
        viewBox="0 0 100 100"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.6 3.5C5.9 13.6-6.6 42.7 3.5 68.4c10.1 25.7 39.2 38.3 64.9 28.1l-3.1-7.9C44 97 19.9 86.6 11.5 65.3c-8.4-21.3 2-45.4 23.3-53.8l-3.2-8z"
        >
          <animatetransform
            attributeName="transform"
            attributeType="XML"
            dur="2s"
            from="0 50 50"
            repeatCount="indefinite"
            to="360 50 50"
            type="rotate"
          />
        </path>
        <path
          d="M42.3 39.6c5.7-4.3 13.9-3.1 18.1 2.7 4.3 5.7 3.1 13.9-2.7 18.1l4.1 5.5c8.8-6.5 10.6-19 4.1-27.7-6.5-8.8-19-10.6-27.7-4.1l4.1 5.5z"
        >
          <animatetransform
            attributeName="transform"
            attributeType="XML"
            dur="1s"
            from="0 50 50"
            repeatCount="indefinite"
            to="-360 50 50"
            type="rotate"
          />
        </path>
        <path
          d="M82 35.7C74.1 18 53.4 10.1 35.7 18S10.1 46.6 18 64.3l7.6-3.4c-6-13.5 0-29.3 13.5-35.3s29.3 0 35.3 13.5l7.6-3.4z"
        >
          <animatetransform
            attributeName="transform"
            attributeType="XML"
            dur="2s"
            from="0 50 50"
            repeatCount="indefinite"
            to="360 50 50"
            type="rotate"
          />
        </path>
      </svg>
      <span
        class="c3 c4"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`Center Aligned Loader should render properly with default props 1`] = `
.c0 {
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 1001;
  position: relative;
}

.c0 > * {
  -webkit-animation: hHgZmL 300ms ease-in;
  animation: hHgZmL 300ms ease-in;
}

<div>
  <div
    class="c0"
  >
    <svg
      class="sc-qPLKk NhgoW"
      fill="none"
      height="51"
      viewBox="0 0 51 51"
      width="51"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="25.5"
        cy="25.5"
        r="24"
        stroke="url(#lg)"
        stroke-width="3"
      >
        <animatetransform
          attributeName="transform"
          attributeType="XML"
          dur="900ms"
          from="0 25.5 25.5"
          repeatCount="indefinite"
          to="360 25.5 25.5"
          type="rotate"
        />
      </circle>
      <defs>
        <lineargradient
          id="lg"
        >
          <stop
            offset="70%"
            stop-color="#EFF2F4"
          />
          <stop
            offset="95%"
            stop-color="#546A7F"
          />
        </lineargradient>
      </defs>
    </svg>
  </div>
</div>
`;
